src/values/ArrayValue.js

```diff
@@ -105,14 +99,70 @@ function evaluatePossibleNestedOptimizedFunctionsAndStoreEffects(
   }
 }
 
+/*
+  We track aliases explicitly, because we currently do not have the primitives to model objects created
+inside of the loop. TODO: Revisit when #2543 and subsequent modeling work
+lands. At that point, instead of of a mayAliasSet, we can return a widened
+abstract value.
+*/
+function modelUnknownPropertyOfSpecializedArray(
+  realm: Realm,
+  array: ArrayValue,
+  possibleNestedOptimizedFunctions: ?PossibleNestedOptimizedFunctions
+): PropertyBinding {
+  let sentinelProperty = {
+    key: undefined,
+    descriptor: new PropertyDescriptor({
+      writable: true,
+      enumerable: true,
+      configurable: true,
+    }),
+    object: array,
+  };
+
+  let mayAliasedObjects: Set<ObjectValue> = new Set();
+
+  if (realm.arrayNestedOptimizedFunctionsEnabled && possibleNestedOptimizedFunctions) {
+    invariant(possibleNestedOptimizedFunctions.length > 0);
```

Comment:
You also need to include objects into `mayAliasedObjects` which the array may already alias.